`guppylang.error.InternalGuppyError`: Unexpected assignment pattern

```py
module = GuppyModule("test")
@guppy(module)
def main() -> None:
    bs = [True]
    bs[0] = bs[0]
```

Running this errors out with
```
guppylang/error.py:101: in pretty_errors_wrapped
    return f(*args, **kwargs)
guppylang/module.py:346: in compile
    self.check()
guppylang/module.py:323: in check
    other_defs = self._check_defs(
guppylang/module.py:295: in _check_defs
    defn.check(parsed_globals) if isinstance(defn, CheckableDef) else defn
guppylang/definition/function.py:106: in check
    cfg = check_global_func_def(self.defined_at, self.ty, globals)
guppylang/checker/func_checker.py:80: in check_global_func_def
    return check_cfg(cfg, inputs, ty.output, generic_params, func_def.name, globals)
guppylang/checker/cfg_checker.py:84: in check_cfg
    checked_cfg.entry_bb = check_bb(
guppylang/checker/cfg_checker.py:200: in check_bb
    checked_stmts = StmtChecker(ctx, bb, return_ty).check_stmts(bb.statements)
guppylang/checker/stmt_checker.py:91: in check_stmts
    return [self.visit(s) for s in stmts]
guppylang/ast_util.py:53: in visit
    return visitor(node, *args, **kwargs)
guppylang/checker/stmt_checker.py:279: in visit_Assign
    node.targets = [self._check_assign(target, node.value, ty)]
../../.local/share/uv/python/cpython-3.12.5-macos-aarch64-none/lib/python3.12/functools.py:944: in _method
    return method.__get__(obj, cls)(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <guppylang.checker.stmt_checker.StmtChecker object at 0x1117337d0>
lhs = <ast.Subscript object at 0x1117aec10>
rhs = <guppylang.nodes.PlaceNode object at 0x111710290>
rhs_ty = OpaqueType(args=[], defn=OpaqueTypeDef(id=DefId(id=3, module=None), name='bool', defined_at=None, description='type', params=[], always_linear=False, to_hugr=<function <lambda> at 0x11115dc60>, bound=None))

    @functools.singledispatchmethod
    def _check_assign(self, lhs: ast.expr, rhs: ast.expr, rhs_ty: Type) -> ast.expr:
        """Helper function to check assignments with patterns."""
>       raise InternalGuppyError("Unexpected assignment pattern")
E       guppylang.error.InternalGuppyError: Unexpected assignment pattern

guppylang/checker/stmt_checker.py:104: InternalGuppyError
```

I'd expect this to emit a nicer error rather than panic internally.

--- a/guppylang/checker/stmt_checker.py
+++ b/guppylang/checker/stmt_checker.py
@@ -150,6 +150,12 @@ class StmtChecker(AstVisitor[BBStatement]):
         return with_loc(lhs, with_type(rhs_ty, PlaceNode(place=place)))
 
     @_check_assign.register
+    def _check_subscript_assign(
+        self, lhs: ast.Subscript, rhs: ast.expr, rhs_ty: Type
+    ) -> AnyUnpack:
+        raise GuppyError(UnsupportedError(lhs, "Subscript assignments"))
+
+    @_check_assign.register
     def _check_tuple_assign(
         self, lhs: ast.Tuple, rhs: ast.expr, rhs_ty: Type
     ) -> AnyUnpack:

--- a/tests/error/array_errors/mutate.err
+++ b/tests/error/array_errors/mutate.err
@@ -1,0 +1,8 @@
+Error: Unsupported (at $FILE:11:4)
+   | 
+ 9 | @guppy(module)
+10 | def main(xs: array[int, 10]) -> None:
+11 |     xs[0] = 1
+   |     ^^^^^ Subscript assignments are not supported
+
+Guppy compilation failed due to 1 previous error

--- a/tests/error/array_errors/mutate.py
+++ b/tests/error/array_errors/mutate.py
@@ -1,0 +1,14 @@
+from guppylang.decorator import guppy
+from guppylang.module import GuppyModule
+from guppylang.std.builtins import array, owned
+
+
+module = GuppyModule("test")
+
+
+@guppy(module)
+def main(xs: array[int, 10]) -> None:
+    xs[0] = 1
+
+
+module.compile()

